lib/ansible/module_utils/oneview.py

```diff
@@ -0,0 +1,782 @@
+                                      HPOneViewValueError,
+                                      HPOneViewResourceNotFound)
+
+    HAS_HPE_ONEVIEW = True
+except ImportError:
+    HAS_HPE_ONEVIEW = False
+
+
+class OneViewModuleBase(object):
+    MSG_CREATED = 'Resource created successfully.'
+    MSG_UPDATED = 'Resource updated successfully.'
+    MSG_DELETED = 'Resource deleted successfully.'
+    MSG_ALREADY_PRESENT = 'Resource is already present.'
+    MSG_ALREADY_ABSENT = 'Resource is already absent.'
+    HPE_ONEVIEW_SDK_REQUIRED = 'HPE OneView Python SDK is required for this module.'
+    FUTURE_PACKAGE_REQUIRED = 'The Future Python package is required for this module.'
+
+    ONEVIEW_COMMON_ARGS = dict(
+        config=dict(required=False, type='str')
+    )
+
+    ONEVIEW_VALIDATE_ETAG_ARGS = dict(
+        validate_etag=dict(
+            required=False,
+            type='bool',
+            default=True)
+    )
+
+    resource_client = None
+
+    def __init__(self, additional_arg_spec=None, validate_etag_support=False):
+        """
+        OneViewModuleBase constructor.
+
+        Args:
+            additional_arg_spec (dict): Additional argument spec definition.
+            validate_etag_support (bool): Enables support to eTag validation.
+        """
+
+        argument_spec = self.__build_argument_spec(additional_arg_spec, validate_etag_support)
+
+        self.module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+
+        self.__check_hpe_oneview_sdk()
+        self.__create_oneview_client()
+
+        self.state = self.module.params.get('state')
+        self.data = self.module.params.get('data')
+
+        # Preload params for get_all - used by facts
+        self.facts_params = self.module.params.get('params') or {}
+
+        # Preload options as dict - used by facts
+        self.options = self.transform_list_to_dict(self.module.params.get('options'))
+
+        self.validate_etag_support = validate_etag_support
+
+    def __build_argument_spec(self, additional_arg_spec, validate_etag_support):
+
+        merged_arg_spec = dict()
+        merged_arg_spec.update(self.ONEVIEW_COMMON_ARGS)
+
+        if validate_etag_support:
+            merged_arg_spec.update(self.ONEVIEW_VALIDATE_ETAG_ARGS)
+
+        if additional_arg_spec:
+            merged_arg_spec.update(additional_arg_spec)
+
+        return merged_arg_spec
+
+    def __check_hpe_oneview_sdk(self):
+        if not HAS_HPE_ONEVIEW:
+            self.module.fail_json(msg=self.HPE_ONEVIEW_SDK_REQUIRED)
+        if not HAS_FUTURE:
+            self.module.fail_json(msg=self.FUTURE_PACKAGE_REQUIRED)
+
+    def __create_oneview_client(self):
+        if not self.module.params['config']:
+            self.oneview_client = OneViewClient.from_environment_variables()
+        else:
+            self.oneview_client = OneViewClient.from_json_file(self.module.params['config'])
+
+    def execute_module(self):
```

Comment:
You could use an abstractbaseclass to enforce the requirement for subclasses to implement execute_module() instead of raising an exception inside of the function.  The advantage of an abc is that it will raise when the class is parsed rather than when the function is called.  https://docs.python.org/2/library/abc.html